src/api/ethereum.js

```diff
@@ -0,0 +1,245 @@
+import {
+  createThirdwebClient, getContract, readContract, prepareContractCall, sendTransaction, waitForReceipt,
+} from 'thirdweb';
+import { getAllWalletsList, injectedProvider } from 'thirdweb/wallets';
+import { ethers5Adapter } from 'thirdweb/adapters/ethers5';
+import { defineChain } from 'thirdweb/chains';
+import { providers } from 'ethers';
+import { resolveMappedPromises } from '../utils/promise';
+
+const getThirdWebContract = (contractAddress) => {
+  const clientId = process.env.REACT_APP_THIRD_WEB_CLIENT_ID;
+  const chainId = process.env.REACT_APP_THIRD_WEB_CHAIN_ID;
+  const rpcUrl = process.env.REACT_APP_THIRD_WEB_RPC_URL;
+  const nativeCurrency = JSON.parse(process.env.REACT_APP_THIRD_WEB_NATIVE_CURRENCY);
+
+  // create the client with your clientId, or secretKey if in a server environment
+  const client = createThirdwebClient({
+    clientId,
+  });
+  const chain = defineChain({
+    id: parseInt(chainId),
+    rpc: rpcUrl,
+    nativeCurrency,
+  });
+  // connect to your contract
+  const contract = getContract({
+    client,
+    chain,
+    address: contractAddress,
+  });
+
+  return {
+    client,
+    contract,
+    chain,
+  };
+};
+
+const resolveOperationFactory = (contractAddress, account) => async (methodName, params = [], wei = 0) => {
+  const { contract, client, chain } = getThirdWebContract(contractAddress);
+
+  const defaultParams = {
+    contract,
+    method: methodName,
+    params,
+  };
+
+  const transaction = await prepareContractCall(wei === 0 ? defaultParams : {
+    ...defaultParams,
+    value: wei,
+  });
+  const adaptedAccount = await ethers5Adapter.signer.fromEthers({
+    signer: account,
+  });
+  const { transactionHash } = await sendTransaction({
+    transaction,
+    account: adaptedAccount,
+  });
+  await waitForReceipt({
+    chain,
+    client,
+    transactionHash,
+    maxBlocksWaitTime: 20,
+  });
+};
+
+const getERC20Operations = (erc20Address, account) => {
+  const resolveOperation = resolveOperationFactory(erc20Address, account);
+  const approve = (spender, value) => resolveOperation(
+    'function approve(address spender, uint256 value) external returns (bool)',
+    [
+      spender, value,
+    ],
+  );
+  return {
+    approve,
+  };
+};
+
+const getTokenStakeOperations = (account, erc20Address) => {
```

Comment:
why do we have getTokenStakeOperations that returns claimRewards and stake in an object instead of getClaimRewards and getStakeOperation separately? Seems like an anti pattern

Reply:
Simplified operations into simple calls to API